hts_expr.c

```diff
@@ -546,34 +689,45 @@ static int eq_expr(hts_filter_t *filt, void *data, hts_expr_sym_func *fn,
 
     str = ws(*end);
 
-    int err = 0;
+    int err = 0, eq_done = 0;
     hts_expr_val_t val = HTS_EXPR_VAL_INIT;
 
     // numeric vs numeric comparison is as expected
     // string vs string comparison is as expected
     // numeric vs string is false
     if (str[0] == '=' && str[1] == '=') {
+        eq_done = 1;
         if ((err = eq_expr(filt, data, fn, str+2, end, &val))) {
             res->is_true = res->d = 0;
         } else {
-            res->is_true = res->d = res->is_str
-                ? (res->s.s && val.s.s ? strcmp(res->s.s, val.s.s)==0 : 0)
-                : !res->is_str && !val.is_str && res->d == val.d;
+            if (!hts_expr_val_exists(res) || !hts_expr_val_exists(&val)) {
+                hts_expr_val_undef(res);
+            } else {
+                res->is_true = res->d = res->is_str
+                    ? (res->s.s && val.s.s ?strcmp(res->s.s, val.s.s)==0 :0)
+                    : !res->is_str && !val.is_str && res->d == val.d;
+            }
         }
         res->is_str = 0;
 
     } else if (str[0] == '!' && str[1] == '=') {
+        eq_done = 1;
         if ((err = eq_expr(filt, data, fn, str+2, end, &val))) {
             res->is_true = res->d = 0;
         } else {
-            res->is_true = res->d = res->is_str
-                ? (res->s.s && val.s.s ? strcmp(res->s.s, val.s.s) != 0 : 1)
-                : res->is_str != val.is_str || res->d != val.d;
+            if (!hts_expr_val_exists(res) || !hts_expr_val_exists(&val)) {
+                hts_expr_val_undef(res);
+            } else {
+                res->is_true = res->d = res->is_str
+                    ? (res->s.s && val.s.s ?strcmp(res->s.s, val.s.s) != 0 :1)
+                    : res->is_str != val.is_str || res->d != val.d;
+            }
         }
         res->is_str = 0;
 
     } else if ((str[0] == '=' && str[1] == '~') ||
                (str[0] == '!' && str[1] == '~')) {
+        eq_done = 1;
```

Comment:
The check already exists with `!res->is_str`.  I could add an additional exists check, but it wouldn't change anything I think.

Non-const string is an interesting one.  You're correct it would fail, eg with `samtools view -e '[AB] =~ [CD]'` with tags AB being a string and CD being a regexp to match it.  Cryptic!  We'd need to cache the string too and check it so we can invalidate the `regcomp` result, but for now I'm inclined to simply document it as only accepting constant strings.